kvdb/interface.go

```diff
@@ -109,6 +109,10 @@ type ExtendedRBucket interface {
 
 	// Prefetch will attempt to prefetch all values under a path.
 	Prefetch(paths ...[]string)
+
+	// ForAll is an optimized version of ForEach with the limitation that no
+	// additional queries can be executed within the callback.
+	ForAll(func(k, v []byte) error) error
```

Comment:
Went with `ForAll` after all. We produced some alternatives, but nothing really captures what it is. So perhaps a more neutral name is better.

Reply:
Something that could be problematic is that even though there's an intention to avoid the misuse of `ForAll` it's not impossible to do so. We could agree this to be a minor issue since we can catch such problems during code reviews. 

Imho the previously suggested and discussed extension to the `DB` interface could be more safe in this regard.

Reply:
Yes, agreed that it could be more safe. On the other hand, a nested transaction will generate an error right away with postgres, so it is unlikely that such an issue wouldn't be detected.

Reply:
Yeah I see your point, still until we have `bolt` and `etcd` implementations we need to thing about those too. Since `ForAll` falls back to `ForEach` such an error is only visible on `postgres`. I think we can implement this behavior with `etcd` so it'd error but not sure about `bolt`. Maybe no need to be super strict just thinking out loud. wdyt @guggero ?

Reply:
The "problem" you're speaking about is that a developer could manually open a new TX within the callback of a `ForAll` call? Or what are we trying to prevent?

Reply:
A new tx isn't allowed in any case, but problem is that within the callback it is also not allowed to do any other query with the same tx.

Reply:
So we can't nest into a sub bucket inside a `ForAll`, nor can't we get an arbitrary key etc.

Reply:
That's correct. But also, if you would get keys inside the callback, it might defeat the whole purpose of it because you're still doing queries per row.

Reply:
I think documentation should be enough to make sure this isn't used incorrectly. Perhaps we should make it a bit more obvious by using the `NOTE: ...` comment pattern.